greenheart/tools/optimization/gc_PoseOptimization.py

```diff
@@ -0,0 +1,505 @@
+from greenheart.tools.optimization.openmdao import TurbineDistanceComponent, BoundaryDistanceComponent
+from hopp.simulation import HoppInterface
+
+class PoseOptimization(object):
+    """_summary_
+
+    Args:
+        object (object): built-in object inheritance
+    """
+    def __init__(self, config: GreenHeartSimulationConfig):
+        self.config = config
+
+        self.nlopt_methods = [
+            "GN_DIRECT",
+            "GN_DIRECT_L",
+            "GN_DIRECT_L_NOSCAL",
+            "GN_ORIG_DIRECT",
+            "GN_ORIG_DIRECT_L",
+            "GN_AGS",
+            "GN_ISRES",
+            "LN_COBYLA",
+            "LD_MMA",
+            "LD_CCSAQ",
+            "LD_SLSQP",
+        ]
+
+        self.scipy_methods = [
+            "SLSQP",
+            "Nelder-Mead",
+            "COBYLA",
+        ]
+
+        self.pyoptsparse_methods = [
+            "SNOPT",
+            "CONMIN",
+            "NSGA2",
+        ]
+
+    def get_number_design_variables(self):
+        """This method counts the number of design variables required given the provided set up and returns the result
+
+        Returns:
+            int: number of design variables
+        """
+        # Determine the number of design variables
+        n_DV = 0
+        
+        if self.config.greenheart_config["opt_options"]["design_variables"]["electrolyzer_rating_kw"]["flag"]:
+            n_DV += 1
+        if self.config.greenheart_config["opt_options"]["design_variables"]["pv_capacity_kw"]["flag"]:
+            n_DV += 1
+        if self.config.greenheart_config["opt_options"]["design_variables"]["wave_capacity_kw"]["flag"]:
+            n_DV += 1
+        if self.config.greenheart_config["opt_options"]["design_variables"]["battery_capacity_kw"]["flag"]:
+            n_DV += 1
+        if self.config.greenheart_config["opt_options"]["design_variables"]["battery_capacity_kwh"]["flag"]:
+            n_DV += 1
+        if self.config.greenheart_config["opt_options"]["design_variables"]["turbine_x"]["flag"]:
+            n_DV += self.config.hopp_config["technologies"]["wind"]["num_turbines"]
+        if self.config.greenheart_config["opt_options"]["design_variables"]["turbine_y"]["flag"]:
+            n_DV += self.config.hopp_config["technologies"]["wind"]["num_turbines"]
+        
+        # Wrap-up at end with multiplier for finite differencing
+        if self.config.greenheart_config["opt_options"]["driver"]["optimization"]["form"] == "central": # TODO this should probably be handled at the MPI point to avoid confusion with n_DV being double what would be expected
+            n_DV *= 2
+
+        return n_DV
+
+    def _get_step_size(self):
+        """If a step size for the driver-level finite differencing is provided, use that step size. Otherwise use a default value.
+
+        Returns:
+            step size (float): step size for optimization
+        """
+
+        return (
```

Comment:
It might be good to rewrite so it checks if either of those values are there or it raises an error with a description of how step size must be set.

Reply:
Good suggestion. Done